Add support for ServiceBinding

yharish991 opened the pull request:
Resolves https://github.com/dekorateio/dekorate/issues/678 https://github.com/dekorateio/dekorate/issues/679 https://github.com/dekorateio/dekorate/issues/680 https://github.com/dekorateio/dekorate/issues/681

iocanel reviewed (changes requested):
There is a copy and paste issue we need to address.
Apart from that the rest are not change requests per-se, but more like discussion topics on how we could possibly eliminate the amount of configuration required.

yharish991 commented on annotations/servicebinding-annotations/pom.xml:
```diff
@@ -0,0 +1,216 @@
+                <filter>
+                  <artifact>*:*</artifact>
+                  <includes>
+                    <include>io/dekorate/component/adapter/**</include>
```
oops, my bad, it was a copy-paste mistake, I pushed a new commit to address this

yharish991 commented on examples/servicebinding-example/src/main/java/io/dekorate/examples/servicebinding/Main.java:
```diff
@@ -0,0 +1,34 @@
+import io.dekorate.servicebinding.annotation.Service;
+import io.dekorate.servicebinding.annotation.ServiceBinding;
+
+@ServiceBinding(name = "servicebinding-demo", services = {
```
`name` parameter is optional here, I just used it in the example to tell that they can even use `name` parameter to customize the ServiceBinding CR name. If the name parameter is not provided, it uses `applicationName` + `-binding` as the CR name. If this is confusing I can rename the parameter to something more meaningful, suggestions are accepted :)

iocanel replied:
Perfect, in this case I would recommend our initial examples to demonstrate the minimum amount of configuration required. And then have either examples or tests demonstrating common use-cases.

FWIW, I find that sometimes we are getting carried away in the examples and demonstrate a lot of options at once. At some point I was asked for a `hello-world` type of example for kubernetes and didn't had one. So, I would at least want a `hello-world` kind of example and then maybe have addtional ones showcasing one feature at a time.

yharish991 replied:
makes sense, i will trim down the example to use minimum configuration

yharish991 commented on annotations/servicebinding-annotations/src/main/java/io/dekorate/servicebinding/annotation/Application.java:
```diff
@@ -0,0 +1,43 @@
+@Retention(RetentionPolicy.SOURCE)
+public @interface Application {
+
+  String group() default "";
```
`@Application` annotation is optional, the developers need not specify those. I too thought of having `BindingPath` directly under `@ServiceBinding`  but then I thought that would deviate the actual servicebinding CR yaml structure and would make it difficult for users to understand. ServiceBinding CR yaml is not too complex, so I thought I will just follow it and make it easy for developers to write the annotations keeping the yaml structure in mind.

Based on your experience writing dekorate annotations, do you think it's a good idea to deviate from the original yaml structure even if it is not too complex? if so I can bring it under `@ServiceBinding` directly.

iocanel replied:
So far, we deviated from the original structure either for techincal reasons, or to keep things simpler. If the use of the annotation is optional, then I don't have a very strong opinion. 
If you feel that it will make it easier for the users to understand, we can pick it and just explain in javadoc, how these values will be poplulated when the annotation is not used.

yharish991 replied:
sometimes I feel it's okay to deviate from the original structure, the reason why I think it's okay is almost 99.99% of the times we don't need to use `@Application` annotation unless you are specifying the `@BindingPath` annotation in the `@Application` because all of the other parameters can be populated by us, if we bring `@BindingPath` directly under `@ServiceBinding` , developers don't have to do this ```application = @Application(bindingPath = @BindingPath(secretPath = "xyz"))``` and need to do this ```bindingPath = @BindingPath(secretPath = "xyz")``` under `@ServiceBinding` which makes it simple

I will move `@BindingPath` directly under `@ServiceBinding`

yharish991 commented:
@iocanel can you review the PR, added `valueFrom` annotation under `customEnvVar`, more tests and documentation

iocanel commented:
@yharish991: regarding the `valueFrom` added to CustomEnvVar, I think that we should not re-invent the wheel and reuse the `@Env`  annotation that is already available in core.

yharish991 commented:
Service Binding Operator made some changes again on their side, they renamed `customEnvVar` to `mappings` https://github.com/redhat-developer/service-binding-operator/pull/802

yharish991 commented:
@iocanel can you re-review my pr? i made some changes as service binding operator made some changes on their side. Here are some major changes they made recently
1. renamed the `customEnvVar` to `mappings` and removed `valueFrom`
2. renamed `envVarPrefx` to `namePrefix`

iocanel commented:
> @iocanel can you re-review my pr? i made some changes as service binding operator made some changes on their side. Here are some major changes they made recently
> 
> 1. renamed the `customEnvVar` to `mappings` and removed `valueFrom`
> 2. renamed `envVarPrefx` to `namePrefix`

Yeah, will try to get this in today!

yharish991 commented:
thanks @iocanel but before we merge this, can we have a quick chat tmrw?

iocanel reviewed (approved):
LGTM, please squash your commits so that we can merge.

yharish991 commented:
closing this PR, will open a new one
